lib/Sema/LookupVisibleDecls.cpp

```diff
@@ -1171,167 +1171,131 @@ static void lookupVisibleMemberDecls(
+          return false;
+      }
+      return true;
+    };
 
-    DC = DC->getParentForLookup();
-  }
+    if (!canAccessInstanceMembers())
+      LS = LS.withOnMetatype();
 
-  if (auto SF = dyn_cast<SourceFile>(DC)) {
-    if (Loc.isValid()) {
-      // Look for local variables in top-level code; normally, the parser
-      // resolves these for us, but it can't do the right thing for
-      // local types.
-      namelookup::FindLocalVal(SM, Loc, Consumer).checkSourceFile(*SF);
-    }
+    auto MemberKind = InnermostTypeDC == DC
+                          ? DeclVisibilityKind::MemberOfCurrentNominal
+                          : DeclVisibilityKind::MemberOfOutsideNominal;
 
-    if (IncludeTopLevel) {
-      auto &cached = SF->getCachedVisibleDecls();
-      if (!cached.empty()) {
-        for (auto result : cached)
-          Consumer.foundDecl(result, DeclVisibilityKind::VisibleAtTopLevel);
-        return;
-      }
-    }
+    lookupVisibleMemberDecls(DC->getSelfTypeInContext(), LookupLoc,
+                             BaseConsumer, DC, LS, MemberKind);
+    return false;
   }
+};
+} // end anonymous namespace
 
-  if (IncludeTopLevel) {
-    using namespace namelookup;
-    SmallVector<ValueDecl *, 0> moduleResults;
-    lookupVisibleDeclsInModule(DC, {}, moduleResults,
-                               NLKind::UnqualifiedLookup,
-                               ResolutionKind::Overloadable,
-                               DC);
-    for (auto result : moduleResults)
+static void lookupVisibleDeclsInModule(const SourceFile *SF,
+                                       VisibleDeclConsumer &Consumer) {
+  auto &cached = SF->getCachedVisibleDecls();
+  if (!cached.empty()) {
+    for (auto result : cached)
       Consumer.foundDecl(result, DeclVisibilityKind::VisibleAtTopLevel);
-
-    if (auto SF = dyn_cast<SourceFile>(DC))
-      SF->cacheVisibleDecls(std::move(moduleResults));
+    return;
   }
+  using namespace namelookup;
+  SmallVector<ValueDecl *, 0> moduleResults;
+  namelookup::lookupVisibleDeclsInModule(SF, {}, moduleResults,
+                                         NLKind::UnqualifiedLookup,
+                                         ResolutionKind::Overloadable, SF);
+  for (auto result : moduleResults)
+    Consumer.foundDecl(result, DeclVisibilityKind::VisibleAtTopLevel);
+
+  SF->cacheVisibleDecls(std::move(moduleResults));
 }
 
-void swift::lookupVisibleDecls(VisibleDeclConsumer &Consumer,
-                               const DeclContext *DC,
-                               bool IncludeTopLevel,
-                               SourceLoc Loc) {
-  if (Loc.isInvalid()) {
-    lookupVisibleDeclsImpl(Consumer, DC, IncludeTopLevel, Loc);
-    return;
+void swift::lookupVisibleDecls(VisibleDeclConsumer &ParentConsumer,
+                               SourceLoc Loc, const DeclContext *DC,
+                               bool IncludeTopLevel) {
+  auto *SF = DC->getParentSourceFile();
+  assert(SF);
+  assert(Loc.isValid());
```

Comment:
Any objections to requiring a valid SourceLoc and SourceFile? We could potentially fall back to a module-scope lookup, but it's not clear to me how useful such a behavior would be (seems like callers ought to use `lookupVisibleDeclsInModule` if that's what they really want), and it seems like all the current clients meet this requirement.

Reply:
No objection. Maybe we should even add
```
assert(DC->isModuleScopeContext());
```
because without `IncludeTopLevel: true`, that lookup doesn't make any sense. If the caller really wants _only_ top-level decls, they usually should use `ModuleDecl::lookupVisibleDecls()`

Reply:
Presumably you mean `assert(!DC->isModuleScopeContext())`? Seems reasonable enough to me

Reply:
Presumably you mean `assert(!DC->isModuleScopeContext());`? Looks like that can be hit for e.g decl attribute completion, and I think seems reasonable to have a module scope context there

Reply:
OK. But just out of curiosity, does `swift::lookupVisibleDecls()` at `SourceFile` context without `IncludeTopLevel` emit anything?

Reply:
No it doesn't include anything